### ISSUE TITLE
Kaldi engine recognizes empty Dictation elements while Dragon engine does not

### ISSUE THREAD

**Issue by caspark** [dictation-toolbox/dragonfly#217]
**Reproduction steps:**

1. Load the following rule into a Dragon grammar

```
class SampleRule(dragonfly.CompoundRule):
    spec = "alpha <text>"
    extras = [
        dragonfly.Dictation(name="text")
    ]

    def _process_recognition(self, node, extras):
        print(f"Sample recognized! node={node} and extras={extras}")
```

2. Say the two utterances `alpha hello` and `alpha`
3. Load the same rule up into Kaldi and say the same utterances

**Observed:** Dragon will recognize `alpha hello` but not `alpha` on its own, because it doesn't allow empty `Dictation` elements to be recognized. Kaldi will recognize both `alpha hello` and `alpha` on its own - but then issues an error like `ERROR:engine:Grammar mygrammar: failed to decode rule SampleRule recognition ('alpha',).` for the latter utterance.

**Expected:** Kaldi's behavior should match Dragon's - don't allow empty Dictations to be recognized.

**Notes:**

* This is mainly a problem if your grammar allows interleaving other commands with dictation.  For example, pretend we have `(alpha <text> | beta)` wrapped in a `dragonfly.Sequence`, and then you say `alpha beat up beta`; Kaldi might recognize that as `alpha beta beta`, which is annoying because it's not what you meant, but also it's annoying because "nothing happens" (not even whatever Action might be associated with `beta`) due to Kaldi recognizing something that Dragonfly doesn't think is possible,
* Not sure if this should be filed here or in https://github.com/daanzu/kaldi-active-grammar - I tried to do some digging to find out but I'm not familiar enough with the details of speech recognition to figure it out so I'm just filing it here. Let me know if you want it moved @daanzu .

**Workaround:** if this is problematic in grammars, you can define a mandatory or optional marker word which ends dictation, e.g. `alpha <text> zulu`, at the cost of having to speak the extra syllables (thanks @dmakarov for pointing this out).

**Comment by daanzu** [comment]
Hmm, I was afraid this could become an issue at some point, but I didn't have a grammar myself where this occurred. I don't think there is an easy fix for this. I have an idea, but it will take a little work. Thanks for the thorough explanation!

By the way, you could try adjusting the weights some and see if that helps, although it won't fix the empty dictation problem. It might help to encourage or discourage Kaldi jumping in or out of a dictation element. Perhaps I need a different default.

**Comment by caspark** [comment]
This is fixed in kaldi-active-grammar-1.4.0 (though note that you need to be using an updated language model such as `daanzu_20200328_1ep-mediumlm` to get the fix, as I believe the fix was to the language model).

Thanks @daanzu !
